_posts/2019-01-02-Deviceid-Reliability.md

```diff
@@ -0,0 +1,90 @@
+
+
+* **Non realistic number of apps**<br/>
+For a fixed amount of time (e.g. 1 day) it is possible to observe the number of apps visited by each deviceID.
+We see, that there are device identifiers, which can be seen on a huge number of different apps during one day.
+These outliers cannot correspond to real human users.
+![num_uniqu_apps](../images/Deviceid-Reliability/num_unique_apps.png)
+In the plot the number of the vistited apps is standardized (the absolute value minus the overall mean divided by the standard deviation), so that the presence of outliers is evident.
+
+
+* **Short period of time**:<br/>
+When we observe a deviceID for a very short period of time (say 1 minute) but we do not register any other occurrence of the deviceID over several days both in the past and in the future, we do not consider these deviceIDs to be worth of further analysis.
+In fact, it is not sure that we have a real person behind it.
+For example a machine which automatically reset the value would produce the same data.
+![one_over_more_days](../images/Deviceid-Reliability/only_1_observations_over_days.png)
+The plot above represents the observation of a deviceID in a 5 days period, both for users which shown a regular activity (d6, d7, d8) and for those which appear only once (d1, d2, d3).
+Every different device has a different y value to avoid overlaps between the points such that the visualization of the activity for each device is facilitated.
+
+
+* **Randomly generated deviceID**:<br/>
+Some app could provide a randomly generated deviceID.
+If this is the case, it is expected that the same deviceID is not observed in any other apps.
+We can think about two ways to generate the values:
+1) a new random deviceID is created for every acces;
+2) a particular transformation of the deviceID is used, thus the provided deviceID for the same real device is always the same but it will not be allined with the one provided by other apps.
+In both cases the fraction of devices shared with other apps should be zero.
+If we plot the histogram of this fraction we can see that there are no apps around zero.
+We can conclude that there are no apps in the data which generate a random deviceID and there is no need to include this selection in the global one.
+![unique_dev_random](../images/Deviceid-Reliability/unique_deviceid_random.png)
+
+
+Of course the same unreliable deviceID can be flagged by more than one pattern.
+However, there is not a single method among those which can be completely substituted by the others.
+Indeed, we include all of them in our deviceID reliability filter (except the randomly generated one since there are no devices which are randomly generated).
+
+**Put everything into production:**<br/>
+We put this quality filter selection into production creating a oozie pipeline which calls the hive queries that implements the described methods.
+The pipeline runs on a daily base such that as we acquire new data we filter out the broken one.
```

Comment:
The pipeline runs on a daily base, acquiring new data and filtering out the unreliable device identifiers.